### PR TITLE
add axiswise granularity to Float8Tensor

### PR DESCRIPTION
**Pull request by vkuzo**
Summary:

This is a copy-paste of https://github.com/pytorch-labs/float8_experimental/pull/352
which never landed.

Test Plan:

```

```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by vkuzo** [comment]
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* (to be filled)

File: test/float8/test_base.py [hunk 1]
```diff
@@ -129,6 +136,111 @@ def test_copy_(self):
         fp8_b.copy_(fp8_a)
         torch.testing.assert_close(fp8_a._data, fp8_b._data)
 
+    @pytest.mark.parametrize("shape", [(8, 16), (4, 8, 16), (2, 4, 8, 16)])
+    @pytest.mark.parametrize("dim_name", ["first", "last"])
```
Comment (drisspg) [inline test/float8/test_base.py:140]: nit: why not just `0`, `-1`

**Review by drisspg — APPROVED** [review]

File: test/float8/test_base.py (hunk 1 above)
Comment (vkuzo) [reply test/float8/test_base.py:140]: will do, that didn't work in the original PR version but should work now

File: torchao/float8/float8_tensor.py [hunk 2]
```diff
@@ -258,11 +263,19 @@ class Float8Tensor(torch.Tensor):
     * `_data`: the underlying e4m3 or e5m2 data
     * `_scale`: the scale used to scale the original fp32 tensor. We multiply
       by scale to go from fp32 range to fp8 range, and divide by scale to go
-      from fp8 range to fp32 range.
+      from fp8 range to fp32 range. Scale is guaranteed to have a shape compatible
+      with `_data`. For example:
+      - if scaling is tensorwise, `_scale` is a scalar tensor
+      - if scaling is axiswise and _data.shape is [3, 5], `_scale` could have
+        shape [1, 5] or [3, 1]. The dim of the non-one entry defines the scaling
```
Comment (vkuzo) [inline torchao/float8/float8_tensor.py:270]: need to remove `The dim of the non-one entry defines the scaling axis` piece as it's not accurate for tensors which have multiple dimensions with size one

**Review by lw — APPROVED** [review]
I skimmed through it, LGTM!

I have one suggestion to have a more streamlined and extendable API, see below

File: test/float8/test_base.py [hunk 3]
```diff
@@ -129,6 +136,105 @@ def test_copy_(self):
...
+    @pytest.mark.parametrize("axiswise_dim", [0, -1])
+    def test_axiswise_dynamic_cast(self, shape, axiswise_dim):
+        a = torch.randn(*shape, dtype=torch.bfloat16)
+        linear_mm_config = LinearMMConfig()
+        a_fp8 = hp_tensor_to_float8_dynamic(
+            a,
+            e4m3_dtype,
+            linear_mm_config,
+            scaling_granularity=ScalingGranularity.AXISWISE,
+            axiswise_dim=axiswise_dim,
```
[the remark this answers is not in the document]
Comment (vkuzo) [reply test/float8/test_base.py:149]: One wrinkle to work through would be that `Float8Tensor` can be of any rank, but operand inputs to `torch._scaled_mm` are required to be of rank 2, to match `torch.mm|torch.addmm`.

I'm definitely open to making this more flexible in the future.  We've been careful to keep `Float8Tensor` and these utility functions out of the public API, to give us the freedom to make these kinds of changes as other scaling types become more important.

[the remark this answers is not in the document]
Comment (vkuzo) [reply test/float8/test_base.py:149]: also, if someone puts up a PR for ^, sgtm!

File: torchao/float8/float8_linear.py [hunk 4]
```diff
@@ -191,15 +188,6 @@ def __init__(self, *args, **kwargs):
         # would be initialized in every iteration.
         self.enable_pre_and_post_forward = self.config.enable_pre_and_post_forward
 
-        # See the comments in config.py for more details of this option.
```
Comment (vkuzo) [inline torchao/float8/float8_linear.py:194]: technically not related to this PR, but making the test logs non-spammy for now and we can add this back in a better way later

**Pull request merged**